### PR TITLE
[RPC] verifytxoutproof returns object including blockhash

### PR DESCRIPTION
**Pull request by instagibbs**
Allows direct use of the proof to get block header related info without additional parsing, and more directly mirrors `gettxoutproof` arguments.

### CODE REVIEW

File: src/rpc/rawtransaction.cpp [hunk 1]
```diff
@@ -258,19 +258,23 @@ UniValue verifytxoutproof(const JSONRPCRequest& request)
...
             "verifytxoutproof \"proof\"\n"
-            "\nVerifies that a proof points to a transaction in a block, returning the transaction it commits to\n"
-            "and throwing an RPC error if the block is not in our best chain\n"
+            "\nVerifies that a proof points to a transaction in a block, returning the blockhash and transactions it commits to\n"
+            ", throwing an RPC error if the block is not in our best chain\n"
             "\nArguments:\n"
             "1. \"proof\"    (string, required) The hex-encoded proof generated by gettxoutproof\n"
-            "\nResult:\n"
-            "[\"txid\"]      (array, strings) The txid(s) which the proof commits to, or empty array if the proof is invalid\n"
+            "\nResult: (or empty if proof is invalid)\n"
```
Comment (kallewoof) [inline src/rpc/rawtransaction.cpp]: Remove `or` maybe? I.e. `Result: (empty if proof is invalid)\n`

**Review by kallewoof — COMMENTED** [review]
utACK d77cded0b1d026f030ee424b89a01cfd6b7b9000

**Comment by instagibbs** [comment]
updated with nit fix

File: src/rpc/rawtransaction.cpp [hunk 2]
```diff
@@ -258,19 +258,23 @@ UniValue verifytxoutproof(const JSONRPCRequest& request)
     if (request.fHelp || request.params.size() != 1)
         throw std::runtime_error(
             "verifytxoutproof \"proof\"\n"
-            "\nVerifies that a proof points to a transaction in a block, returning the transaction it commits to\n"
-            "and throwing an RPC error if the block is not in our best chain\n"
+            "\nVerifies that a proof points to a transaction in a block, returning the blockhash and transactions it commits to\n"
+            ", throwing an RPC error if the block is not in our best chain\n"
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Nit, `,` is the above line?

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -258,19 +258,23 @@ UniValue verifytxoutproof(const JSONRPCRequest& request)
     if (request.fHelp || request.params.size() != 1)
         throw std::runtime_error(
             "verifytxoutproof \"proof\"\n"
-            "\nVerifies that a proof points to a transaction in a block, returning the transaction it commits to\n"
-            "and throwing an RPC error if the block is not in our best chain\n"
+            "\nVerifies that a proof points to a transaction in a block, returning the blockhash and transactions it commits to\n"
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: `... the block hash and ...`

File: src/rpc/rawtransaction.cpp [hunk 4]
```diff
@@ -258,19 +258,23 @@ UniValue verifytxoutproof(const JSONRPCRequest& request)
...
+            ", throwing an RPC error if the block is not in our best chain\n"
             "\nArguments:\n"
             "1. \"proof\"    (string, required) The hex-encoded proof generated by gettxoutproof\n"
-            "\nResult:\n"
-            "[\"txid\"]      (array, strings) The txid(s) which the proof commits to, or empty array if the proof is invalid\n"
+            "\nResult: (empty if proof is invalid)\n"
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: This is breaking change as the result is now an object instead of an array, needs release note.

File: src/rpc/rawtransaction.cpp [hunk 5]
```diff
@@ -283,7 +287,10 @@ UniValue verifytxoutproof(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found in chain");
 
     for (const uint256& hash : vMatch)
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Nit, add curly braces.

**Review by promag — COMMENTED** [review]
Concept ACK.

File: src/rpc/rawtransaction.cpp (hunk 2 above)
Comment (kallewoof) [reply src/rpc/rawtransaction.cpp]: I was gonna mention that but chose not to. Agree though, I think that would look better.

**Comment by instagibbs** [comment]
fixed @promag nits, curly-bracketed the surrounding statements since I'm touching it already

File: src/rpc/rawtransaction.cpp [hunk 6]
```diff
@@ -258,32 +258,42 @@ UniValue verifytxoutproof(const JSONRPCRequest& request)
...
+    }
+
+    for (const uint256& hash : vMatch) {
+        txids.push_back(hash.GetHex());
+    }
 
-    for (const uint256& hash : vMatch)
-        res.push_back(hash.GetHex());
+    res.push_back(Pair("txids", txids));
+    res.push_back(Pair("blockhash", merkleBlock.header.GetHash().GetHex()));
```
Comment (promag) [inline src/rpc/rawtransaction.cpp]: Push `blockhash` first.

Comment (kallewoof) [reply src/rpc/rawtransaction.cpp]: I'm curious about the reason for this request. Alphabetical ordering?

Comment (promag) [reply src/rpc/rawtransaction.cpp]: Also matches the documentation above.

**Comment by instagibbs** [comment]
rebased, re-added legacy behavior under deprecation

**Comment by instagibbs** [comment]
ping @kallewoof

File: src/rpc/rawtransaction.cpp [hunk 7]
```diff
@@ -294,6 +294,8 @@ UniValue verifytxoutproof(const JSONRPCRequest& request)
             "    \"blockhash\"   (string) The blockhash the included proof commits to\n"
             "    [\"txid\"]      (array, strings) The txid(s) which the proof commits to\n"
             "}\n"
+            "\nResult: (DEPRECATED. To see this result in v0.17 instead, please start bitcoind with -deprecatedrpc=verifytxoutproof. Empty if proof is invalid)\n"
```
Comment (kallewoof) [inline src/rpc/rawtransaction.cpp]: I'm not sure about the `in v0.17 instead` wording -- what about when bitcoin hits v0.18?

**Review by kallewoof — COMMENTED** [review]
re-utACK a3716960de5555572408efdab4eba75484c8c1e3

File: src/rpc/rawtransaction.cpp (hunk 7 above)
Comment (promag) [reply src/rpc/rawtransaction.cpp]: I thought the same, but there are other occurrences in the code. I would remove the version though.

**Comment by promag** [comment]
@instagibbs See https://github.com/bitcoin/bitcoin/pull/11872/files#diff-ef76fd6674f07db88c3422fdbf0bcf9fR65 for release note example.

File: src/rpc/rawtransaction.cpp (hunk 7 above)
Comment (instagibbs) [reply src/rpc/rawtransaction.cpp]: I'm a bit confused, if this is merged now, it definitely would be the way to see the old text for 0.17. Now, perhaps the note should also say when it will be removed?

Comment (instagibbs) [reply src/rpc/rawtransaction.cpp]: Pushed a new wording, hopefully clearer and more pertinent to the user.

**Comment by instagibbs** [comment]
fixed up deprecation message, added release notes

**Review by promag — COMMENTED** [review]
Could add a test for the empty response `{}`, at least I don't see it?

File: doc/release-notes.md [hunk 8]
```diff
@@ -62,6 +62,7 @@ RPC changes
 ### Low-level changes
 
 - The `fundrawtransaction` rpc will reject the previously deprecated `reserveChangeKey` option.
+- The `verifytxoutproof` rpc will output an object instead of an array unless `-deprecatedrpc=verifytxoutproof` is set
```
Comment (promag) [inline doc/release-notes.md]: Nit, missing period, sorry.

**Comment by instagibbs** [comment]
fixed up release note nit. I think the only way to return `{}` is to have a non-matching hash header... too lazy to construct one in paste it in.

**Comment by promag** [comment]
@instagibbs here is one
```
verifytxoutproof 0000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
{
}
```
Maybe it should throw an error in this case?

**Review by kallewoof — APPROVED** [review]
utACK - looks good

**Comment by instagibbs** [comment]
@promag yeah I'll catch the deprecated case of `[]` and throw an error for new behavior

**Comment by instagibbs** [comment]
done

**Comment by MarcoFalke** [comment]
Needs rebase to fix travis (sorry)

**Comment by instagibbs** [comment]
rebased

**Review by kallewoof — APPROVED** [review]
utACK ec1695eedc545545e5273e19f1e1b80aaebecf28

**Comment by promag** [comment]
utACK ec1695e.

**Comment by MarcoFalke** [comment]
Needs rebase (after merge of `LookupBlockIndex`)

**Comment by instagibbs** [comment]
rebased

**Comment by kallewoof** [comment]
@instagibbs Travis is timing out. I tried restarting the two jobs that timed out but it still fails. Not sure what's up.

**Comment by promag** [comment]
master should be fixed after #12681.

**Comment by kallewoof** [comment]
@promag I think that's unrelated (and I don't think @instagibbs merged the commit that triggered the error that was fixed).

File: src/rpc/rawtransaction.cpp [hunk 9]
```diff
@@ -285,24 +285,35 @@ UniValue verifytxoutproof(const JSONRPCRequest& request)
...
 
     std::vector<uint256> vMatch;
     std::vector<unsigned int> vIndex;
-    if (merkleBlock.txn.ExtractMatches(vMatch, vIndex) != merkleBlock.header.hashMerkleRoot)
-        return res;
+    if (merkleBlock.txn.ExtractMatches(vMatch, vIndex) != merkleBlock.header.hashMerkleRoot) {
+        if (IsDeprecatedRPCEnabled("verifytxoutproof")) {
+            return txids;
+        } else {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "No valid transaction commitments are found in the proof");
```
Comment (sipa) [inline src/rpc/rawtransaction.cpp]: Is the RPC user at fault when this happens? If not, I don't think this should be an RPC error.

Comment (instagibbs) [reply src/rpc/rawtransaction.cpp]: We already error out when the block is not found in the chain, which is just as much the users' fault.

I'll leave this up to others to chime in.

**Comment by kallewoof** [comment]
re-utACK

File: src/rpc/rawtransaction.cpp (hunk 9 above)
Comment (sipa) [reply src/rpc/rawtransaction.cpp]: Ok, fair. Let's keep it this way then.

**Comment by instagibbs** [comment]
rebased due to release notes conflict

**Comment by instagibbs** [comment]
rebased again...

**Comment by instagibbs** [comment]
rebased again for yet another release notes conflict
